src/node_http_parser.cc

```diff
@@ -751,21 +735,39 @@ class Parser : public AsyncWrap {
   StreamResource::Callback<StreamResource::AllocCb> prev_alloc_cb_;
   StreamResource::Callback<StreamResource::ReadCb> prev_read_cb_;
   int refcount_ = 1;
+
+  // These are helper functions for filling `http_parser_settings`, which turn
+  // a member function of Parser into a C-style HTTP parser callback.
+  template <int (Parser::* Member)()>
```

Comment:
It's possible to make the template more generic and avoid code duplication like this:

```c++
template <typename Parser, Parser> struct Proxy;
template <typename Parser, typename ...Args, int (Parser::*Member)(Args...)>
struct Proxy<int (Parser::*)(Args...), Member> {
  static int Raw(http_parser* p, Args ... args) {
    Parser* parser = ContainerOf(&Parser::parser_, p);
    return (parser->*Member)(std::forward<Args>(args)...);
  }
};

typedef int(Parser::*Call)();
typedef int(Parser::*DataCall)(const char* at, size_t length);

const struct http_parser_settings Parser::settings = {
  Proxy<Call, &Parser::on_message_begin>::Raw,
  Proxy<DataCall, &Parser::on_url>::Raw,
  ...
}
```

Reply:
Wow! I tried going into that direction but couldn’t get it to compile 😄 Would you prefer your approach? I feel like the current version is already a lot to digest, but your approach is kind of *cool*…

Reply:
I personally prefer DRYing code even it involves a bit more template magic..but it's totally fine that you keep it this way.

Reply:
@joyeecheung Okay, done! Kinda still can’t believe it works. 😄